barretenberg/cpp/src/barretenberg/stdlib_circuit_builders/op_queue/ecc_ops_table.hpp

```diff
@@ -0,0 +1,142 @@
+#pragma once
+
+#include "barretenberg/ecc/curves/bn254/bn254.hpp"
+#include "barretenberg/eccvm/eccvm_builder_types.hpp"
+#include "barretenberg/stdlib_circuit_builders/op_queue/ecc_op_queue.hpp"
+namespace bb {
+
+/**
+ * @brief A table of ECC operations
+ * @details The table is constructed via concatenation of subtables of ECC operations. The table concatentation protocol
+ * (Merge protocol) requires that the concatenation be achieved via PRE-pending successive tables. To avoid the need for
+ * expensive memory reallocations associated with physically prepending, the subtables are stored in a linked list that
+ * can be traversed to reconstruct the columns of the aggregate tables as needed (e.g. in corresponding polynomials).
+ *
+ * @tparam OpFormat Format of the ECC operations stored in the table
+ */
+template <typename OpFormat> class EccOpsTable {
+    using Curve = curve::BN254;
+    using Fr = Curve::ScalarField;
+
+  protected:
+    // A segment of the table. (In practice, each subtable stores the operations of a single circuit)
+    struct Subtable {
+        std::vector<OpFormat> data;
+        std::unique_ptr<Subtable> prev; // the previously constructed subtable in the linked list
+
+        Subtable(size_t size_hint = 0) { data.reserve(size_hint); }
+
+        size_t size() const { return data.size(); }
+    };
+
+    std::unique_ptr<Subtable> current_subtable = nullptr; // the subtable to which new ops are appended
+
+  public:
+    size_t size() const
+    {
+        size_t total_size = 0;
+        for (auto* subtable = current_subtable.get(); subtable != nullptr; subtable = subtable->prev.get()) {
+            total_size += subtable->size();
+        }
+        return total_size;
+    }
+
+    // An iterator to allow range based iteration over the full table
+    class Iterator {
+        Subtable* subtable = nullptr;
+        size_t index = 0;
+        size_t subtable_size = 0;
+
+      public:
+        Iterator() = default;
+        Iterator(Subtable* sub, size_t idx)
+            : subtable(sub)
+            , index(idx)
+            , subtable_size(sub ? sub->size() : 0)
+        {}
```

Comment:
Nit: could get away without storing size

Reply:
yeah I did this to save on the size() call every time in a hot loop but maybe thats overkill. I would def prefer just calling size()

Reply:
Yeah so what's going to happen:
- we are reading a pointer, offsetting it (basically free), then reading a field
vs
- we are reading a field
the pointer we are dereferencing will be in cache, so it's a trivial difference

Reply:
We are also returning a rather big struct, which, fair if we are reading one field from it might be a hot loop, but if we are doing anything with a field the loop overhead is not that big. Comes down to how many times we will be iterating during a proving lifetime, I guess. Could be worth the cycles potentially